File: Graphs/BFS-DFS/number-of-islands.py
from collections import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:


        def dfs(grid, row, col):

            if row < 0 or col < 0 or row >= m or col >= n or grid[row][col] == '2' or grid[row][col] == '0':
                return
            
            grid[row][col] = '2'

            dfs(grid, row - 1, col)
            dfs(grid, row + 1, col)
            dfs(grid, row, col - 1)
            dfs(grid, row, col + 1)

            return

        count = 0

        m = len(grid)
        n = len(grid[0])

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1' :
                    dfs(grid, i, j)
                    count += 1
        
        return count

File: Graphs/BFS-DFS/number-of-islands.cpp
#include<vector>

using namespace std;

class Solution {
    public:
        int numIslands(vector<vector<char>>& grid) {
    
            int c = 0;
    
            int m = grid.size(), n = grid[0].size();
    
            for(int i = 0; i < m; i++){
                for(int j = 0; j < n; j++){
    
                    if(grid[i][j] == '1'){
                        dfs(grid, i, j);
                        c++;
                    }
    
                }
            }
    
            return c;
            
        }
    
        void dfs(vector<vector<char>> &grid, int r, int c){
    
            if( r < 0 || c < 0 || r >= grid.size() || c >= grid[0].size() || grid[r][c] == '0') return;
    
    
            grid[r][c] = '0';
    
            dfs(grid, r + 1, c);
            dfs(grid, r - 1, c);
            dfs(grid, r, c - 1);
            dfs(grid, r, c + 1);
    
    
        }
    };